File: src/gui/SettingsDialog.cpp
// SettingsDialog.cpp — Preferences dialog.

#include "SettingsDialog.h"

#include <QComboBox>
#include <QDialogButtonBox>
#include <QFormLayout>
#include <QSettings>
#include <QSpinBox>
#include <array>

namespace {
// name, half-life (seconds)
const std::array<std::pair<const char*, double>, 7> ISOTOPES = {{
    {"F-18  (6586 s)",   6586.2},
    {"C-11  (1223 s)",   1223.4},
    {"N-13  (598 s)",     597.9},
    {"O-15  (122 s)",     122.24},
    {"Ga-68 (4062 s)",   4062.0},
    {"Cu-64 (45720 s)", 45720.0},
    {"Zr-89 (282240 s)",282240.0},
}};
}

SettingsDialog::SettingsDialog(QWidget* parent) : QDialog(parent)
{
    setWindowTitle("Preferences");
    QSettings s("ROISA", "ROISA");
    auto* form = new QFormLayout(this);

    m_brush = new QSpinBox; m_brush->setRange(1, 50);
    m_brush->setValue(s.value("brushRadius", 3).toInt());
    form->addRow("Default brush radius", m_brush);

    m_cmap = new QComboBox; m_cmap->addItems({"Gray", "Hot", "Cool", "Viridis"});
    m_cmap->setCurrentIndex(s.value("colormap", 0).toInt());
    form->addRow("Default colormap", m_cmap);

    m_iso = new QComboBox;
    for (const auto& it : ISOTOPES) m_iso->addItem(it.first);
    m_iso->setCurrentIndex(s.value("isotope", 0).toInt());
    form->addRow("PET isotope", m_iso);

    auto* bb = new QDialogButtonBox(
        QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
    connect(bb, &QDialogButtonBox::accepted, this, [this]{ persist(); accept(); });
    connect(bb, &QDialogButtonBox::rejected, this, &QDialog::reject);
    form->addRow(bb);
}

void SettingsDialog::persist()
{
    QSettings s("ROISA", "ROISA");
    s.setValue("brushRadius", m_brush->value());
    s.setValue("colormap",    m_cmap->currentIndex());
    s.setValue("isotope",     m_iso->currentIndex());
}

int    SettingsDialog::brushRadius() const { return m_brush->value(); }
int    SettingsDialog::colormap()    const { return m_cmap->currentIndex(); }
double SettingsDialog::halfLifeS()   const { return ISOTOPES[m_iso->currentIndex()].second; }

File: roisa/gui/settings_dialog.py
"""
settings_dialog.py — Preferences dialog (default brush, colormap, isotope).

Values are persisted in QSettings and returned to the caller, which applies
them to the tool panel.
"""

from __future__ import annotations

from PyQt6.QtCore import QSettings
from PyQt6.QtWidgets import (
    QComboBox, QDialog, QDialogButtonBox, QFormLayout, QSpinBox,
)

# Common PET isotopes and their half-lives (seconds)
ISOTOPES = [
    ("F-18  (6586 s)",  6586.2),
    ("C-11  (1223 s)",  1223.4),
    ("N-13  (598 s)",    597.9),
    ("O-15  (122 s)",    122.24),
    ("Ga-68 (4062 s)",  4062.0),
    ("Cu-64 (45720 s)", 45720.0),
    ("Zr-89 (282240 s)",282240.0),
]


class SettingsDialog(QDialog):
    def __init__(self, parent=None) -> None:
        super().__init__(parent)
        self.setWindowTitle("Preferences")
        self._s = QSettings("ROISA", "ROISA")
        form = QFormLayout(self)

        self._brush = QSpinBox(); self._brush.setRange(1, 50)
        self._brush.setValue(int(self._s.value("brushRadius", 3)))
        form.addRow("Default brush radius", self._brush)

        self._cmap = QComboBox(); self._cmap.addItems(["Gray", "Hot", "Cool", "Viridis"])
        self._cmap.setCurrentIndex(int(self._s.value("colormap", 0)))
        form.addRow("Default colormap", self._cmap)

        self._iso = QComboBox()
        for name, _ in ISOTOPES:
            self._iso.addItem(name)
        self._iso.setCurrentIndex(int(self._s.value("isotope", 0)))
        form.addRow("PET isotope", self._iso)

        bb = QDialogButtonBox(
            QDialogButtonBox.StandardButton.Ok | QDialogButtonBox.StandardButton.Cancel)
        bb.accepted.connect(self._accept)
        bb.rejected.connect(self.reject)
        form.addRow(bb)

    def _accept(self) -> None:
        self._s.setValue("brushRadius", self._brush.value())
        self._s.setValue("colormap",    self._cmap.currentIndex())
        self._s.setValue("isotope",     self._iso.currentIndex())
        self.accept()

    def values(self) -> dict:
        return {
            "brushRadius": self._brush.value(),
            "colormap":    self._cmap.currentIndex(),
            "halfLifeS":   ISOTOPES[self._iso.currentIndex()][1],
        }
